### PR TITLE
GPU support for NNIAP: energies

### PR DESCRIPTION
**Pull request by Dariyan-Khan**
GPU support for NNIAP module (but relies on froce descriptors being supplied)

### CODE REVIEW

File: examples/ACE/utils/NNIAP.jl [hunk 1]
```diff
@@ -42,22 +76,73 @@ end
...
     fs, fs_pred = get_all_forces(ds), get_all_forces(ds, nniap)
     return w_e * Flux.mse(es_pred, es) + w_f * Flux.mse(fs_pred, fs)
 end
 
 
+function loss(nn, iap,  batch, true_energy, local_descriptors, _device, w_e=1, w_f=1)
+    # nniap = NNIAP(nn, iap)
+    es_pred = sum(sum(nn(local_descriptors)))
+    fs, fs_pred = get_all_forces(batch), get_all_forces(batch, nn , local_descriptors, _device)
+    return w_e * Flux.mse(es_pred, true_energy) + w_f * Flux.mse(fs_pred, fs)
```
[the remark this answers is not in the document]
Comment (emmanuellujan) [reply examples/ACE/utils/NNIAP.jl:105]: - For fitting the energies, this loss function requires: true energies, a matrix of local descriptors, and the nn model. 
- For fitting the forces, this loss function requieres: true forces, a matrix of force descriptors, and the nn model. I am not sure what is the role of the `batch` here, probably this should be replaced with the matrix of force descriptors.
- All these variables are located in the gpu before the loss function is called, so `_device` is not needed.

File: examples/ACE/utils/NNIAP.jl [hunk 2]
```diff
@@ -42,22 +76,73 @@ end
...
     nniap = NNIAP(nn, iap)
     es, es_pred = get_all_energies(ds), get_all_energies(ds, nniap)
     fs, fs_pred = get_all_forces(ds), get_all_forces(ds, nniap)
     return w_e * Flux.mse(es_pred, es) + w_f * Flux.mse(fs_pred, fs)
 end
 
 
+function loss(nn, iap,  batch, true_energy, local_descriptors, _device, w_e=1, w_f=1)
+    # nniap = NNIAP(nn, iap)
+    es_pred = sum(sum(nn(local_descriptors)))
```
Comment (emmanuellujan) [inline examples/ACE/utils/NNIAP.jl:103]: It is using the definition of the potential energy based on the global descriptors, please use the definition that uses the local descriptors.

File: examples/ACE/utils/NNIAP.jl [hunk 3]
```diff
@@ -42,22 +76,73 @@ end
...
+    fs, fs_pred = get_all_forces(batch), get_all_forces(batch, nn , local_descriptors, _device)
+    return w_e * Flux.mse(es_pred, true_energy) + w_f * Flux.mse(fs_pred, fs)
+end
+
+
 # Auxiliary functions ##########################################################
+function get_all_energies(ds::DataSet, nniap::NNIAP, _device)
```
Comment (emmanuellujan) [inline examples/ACE/utils/NNIAP.jl:110]: It is not necessary to add `_device` to this function, the data transfer to the GPU will be handled by the learn function.

File: examples/ACE/utils/NNIAP.jl [hunk 4]
```diff
@@ -42,22 +76,73 @@ end
...
+    return [potential_energy(ds[c], nniap, _device) for c in 1:length(ds)]
+end
+
+
 function get_all_energies(ds::DataSet, nniap::NNIAP)
     return [potential_energy(ds[c], nniap) for c in 1:length(ds)]
 end
 
+function get_all_forces(ds::DataSet, nniap::NNIAP, _device)
```
Comment (emmanuellujan) [inline examples/ACE/utils/NNIAP.jl:119]: It is not necessary to add `_device` to this function, the data transfer to the GPU will be handled by the learn function.

File: examples/ACE/utils/NNIAP.jl [hunk 5]
```diff
@@ -42,22 +76,73 @@ end
...
+    return reduce(vcat,reduce(vcat,[force(ds[c], nniap, _device) for c in 1:length(ds)]))
+end
+
+function get_all_forces(ds::DataSet, nniap::NNIAP, local_descriptors) # new
+    return reduce(vcat,reduce(vcat,[force(ds[c], nniap, local_descriptors, _device) for c in 1:length(ds)]))
+end
+
```
Comment (emmanuellujan) [inline examples/ACE/utils/NNIAP.jl:126]: The forces are not computed using the local descriptors, but the force descriptors. Let's fit energies for now, then we can tackle the forces. Also, it is not necessary to add `_device` to this function, the data transfer to the GPU will be handled by the learn function.

File: examples/ACE/utils/NNIAP.jl [hunk 6]
```diff
@@ -42,22 +76,73 @@ end
...
+function get_all_forces(ds::DataSet, nn, local_descriptors, _device) # new
+    return reduce(vcat,reduce(vcat,[force(ds[c], nn, local_descriptors) for c in 1:length(ds)]))
+end
+
```
Comment (emmanuellujan) [inline examples/ACE/utils/NNIAP.jl:130]: Same as comment above.

File: examples/ACE/utils/NNIAP.jl [hunk 7]
```diff
@@ -78,6 +163,52 @@ function learn!(nniap, ds, opt::Flux.Optimise.AbstractOptimiser, epochs, loss, w
     end
 end
 
+function learn!(nniap, ds, opt::Flux.Optimise.AbstractOptimiser, epochs, loss, w_e, w_f,_device)
```
Comment (emmanuellujan) [inline examples/ACE/utils/NNIAP.jl:166]: This function is not needed due to the new approach below.

File: examples/ACE/utils/NNIAP.jl [hunk 8]
```diff
@@ -78,6 +163,52 @@ function learn!(nniap, ds, opt::Flux.Optimise.AbstractOptimiser, epochs, loss, w
...
+        grads = ∇gpu_loss(nniap.nn, nniap.iap, ds, w_e, w_f)
+        Flux.update!(optim, nniap.nn, grads[1])
+        # Logging
+        curr_loss = gpu_loss(nniap.nn, nniap.iap, ds, w_e, w_f)
+        push!(losses, curr_loss)
+        println(curr_loss)
+    end
+end
+
+function learn!(nn, iap, ds, opt::Flux.Optimise.AbstractOptimiser, epochs, loss, w_e, w_f, true_energy, local_descriptors, _device=gpu, n_batches=1) # new
```
Comment (emmanuellujan) [inline examples/ACE/utils/NNIAP.jl:189]: Use this interface like this:

`learn!(nace, ds, opt::Flux.Optimise.AbstractOptimiser, n_epochs, n_batches = 1, loss, w_e, w_f, device = gpu)`

File: examples/ACE/utils/NNIAP.jl [hunk 9]
```diff
@@ -78,6 +163,52 @@ function learn!(nniap, ds, opt::Flux.Optimise.AbstractOptimiser, epochs, loss, w
...
+    nniap = NNIAP(nn, iap)
+    optim = Flux.setup(opt, nniap.nn) |> _device
+    ∇loss(nn, iap, batch, true_energy, local_descriptors, _device, w_e, w_f) = gradient((nn) -> loss(nn, iap, batch, true_energy, local_descriptors, _device, w_e, w_f), nn)
```
Comment (emmanuellujan) [inline examples/ACE/utils/NNIAP.jl:192]: Change the interfaces of the loss function according to the above comments.

File: examples/ACE/utils/NNIAP.jl [hunk 10]
```diff
@@ -78,6 +163,52 @@ function learn!(nniap, ds, opt::Flux.Optimise.AbstractOptimiser, epochs, loss, w
...
+    losses = []
+    batch_lists = batch_and_shuffle(collect(1:length(ds)), n_batches)
+    batch_list_len = length(batch_lists)
+    for epoch in 1:epochs
+        true_energy = rand(Float32, 100) |> _device
```
Comment (emmanuellujan) [inline examples/ACE/utils/NNIAP.jl:197]: `true_energy` has to be extracted from the current batch of atomic configurations. The true energy type is Float64. They need to be casted to Float32. Then, they need to be allocated into the GPU.

File: examples/ACE/utils/NNIAP.jl [hunk 11]
```diff
@@ -78,6 +163,52 @@ function learn!(nniap, ds, opt::Flux.Optimise.AbstractOptimiser, epochs, loss, w
...
+        local_descriptors = [rand(Float32, 26) for _ in 1:96]
```
Comment (emmanuellujan) [inline examples/ACE/utils/NNIAP.jl:198]: `local_descriptors`  need to be extracted from the current batch of atomic configurations. The local_descriptors type is Float32, so it is not necessary to cast them.

File: examples/ACE/utils/NNIAP.jl [hunk 12]
```diff
@@ -78,6 +163,52 @@ function learn!(nniap, ds, opt::Flux.Optimise.AbstractOptimiser, epochs, loss, w
...
+        local_descriptors = reduce(hcat, local_descriptors) |> _device
```
Comment (emmanuellujan) [inline examples/ACE/utils/NNIAP.jl:199]: Good, the `local_descriptors` are stacked and transferred into the GPU.

File: examples/ACE/utils/NNIAP.jl [hunk 13]
```diff
@@ -78,6 +163,52 @@ function learn!(nniap, ds, opt::Flux.Optimise.AbstractOptimiser, epochs, loss, w
...
+        batch_index = mod(epoch, batch_list_len) + 1 
+        ds_batch = ds[batch_lists[batch_index]]
+        # Compute gradient with current parameters and update model
+        grads = ∇loss(nniap.nn, nniap.nn, ds_batch, true_energy, local_descriptors, _device, w_e, w_f)
```
Comment (emmanuellujan) [inline examples/ACE/utils/NNIAP.jl:203]: Remember that the loss functions interface should change. See comments above.

**Review by emmanuellujan — COMMENTED** [review]
I left several comments, mainly concerned with cleaning up old code, and some modifications to the new code. When the energy training is working for GPU we can merge the PR. In another PR the GPU calculation of the forces can be addressed.

File: examples/ACE/utils/NNIAP.jl [hunk 14]
```diff
@@ -42,22 +62,60 @@ end
...
 
 
+function loss(nn, iap,  batch, true_energy, local_descriptors, w_e=1, w_f=1)
+    # nniap = NNIAP(nn, iap)
+    es_pred = sum(sum(nn.(local_descriptors)))
+    # fs = get_all_forces(batch)
+    # fs_pred = get_all_forces(batch, nn , local_descriptors, _device)
+    return w_e * Flux.mse(es_pred, true_energy)  # +   w_f * Flux.mse(fs_pred, fs)
+end
+
```
Comment (emmanuellujan) [inline examples/ACE/utils/NNIAP.jl:80]: `e_pred` calculation needs to be fixed:
 
`local_descriptors` is a column-major ordered matrix containing the local descriptors of a set of configurations. E.g., suppose the first configuration c1 has N atoms, and the second configuration c2 has M atoms, then:
`local_descriptors = { ld_atom1_c1, ld_atom2_c1, ..., ld_atomN_c1, ld_atom1_c2, ... ld_atomM_c2, ... }`

where `ld_atomi_cj `is a column vector. In the function `learn!` these vectors where stacked into a matrix and transfered to the GPU.

`e_pred` is a vector of predicted energies associated to a particular batch of configurations (the batch was defined in the function `learn!`). 

The predicted energy of a configuration c (that is, one of the elements of `e_pred`) is defined as:
`e_pred_c = sum_i nn(local_descriptors_i)`
where i represents the atoms of the configuration c

The code `sum(sum(nn.(local_descriptors)))` does not implement the definition above. 

An option is to first apply the nn model to the matrix of local descriptors (`nn.(local_descriptors)`). Them, split the result (that is a CuArray) with respect to the number of atoms of each configuration in the local descriptor matrix (probably you need a new vector with this information). Finally, sum each part to obtain each predicted energy.

File: examples/ACE/utils/NNIAP.jl (hunk 3 above)
Comment (emmanuellujan) [reply examples/ACE/utils/NNIAP.jl:83]: This function `get_all_energies(ds::DataSet, nniap::NNIAP, _device)` is not needed anymore.

File: examples/ACE/utils/NNIAP.jl (hunk 4 above)
Comment (emmanuellujan) [reply examples/ACE/utils/NNIAP.jl:92]: This function `get_all_forces(ds::DataSet, nniap::NNIAP, _device)` is not needed anymore.

File: examples/ACE/utils/NNIAP.jl [hunk 15]
```diff
@@ -72,12 +130,45 @@ function learn!(nniap, ds, opt::Flux.Optimise.AbstractOptimiser, epochs, loss, w
         grads = ∇loss(nniap.nn, nniap.iap, ds, w_e, w_f)
         Flux.update!(optim, nniap.nn, grads[1])
         # Logging
-        curr_loss = loss(nniap.nn, nniap.iap, ds, w_e, w_f)
+        loss_time = @elapsed curr_loss = loss(nniap.nn, nniap.iap, ds, w_e, w_f)
+        println("loss_time", loss_time)
         push!(losses, curr_loss)
```
Comment (emmanuellujan) [inline examples/ACE/utils/NNIAP.jl:135]: To benchmark performance, use `@benchmark` over the function `learn!`, in serial and parallel, using the same parameters (e.g. number of epochs, etc). If you only measure energy training, redefine the serial loss function to only consider energies (just comment the lines related to force calculation). Please, do not commit the benchmarks to this PR.

`@elapsed` is not necessary here, please remove it.

File: examples/ACE/utils/NNIAP.jl [hunk 16]
```diff
@@ -72,12 +130,45 @@ function learn!(nniap, ds, opt::Flux.Optimise.AbstractOptimiser, epochs, loss, w
...
         println(curr_loss)
     end
 end
 
+function learn!(nace, ds_train, opt, n_epochs, n_batches, loss, w_e, w_f, _device)
+    nn = nace.nn |> _device
+    iap = nace.iap |> _device
```
Comment (emmanuellujan) [inline examples/ACE/utils/NNIAP.jl:142]: I don't think it is necessary to transfer the iap to the GPU. Is it?

File: examples/ACE/utils/NNIAP.jl [hunk 17]
```diff
@@ -72,12 +130,45 @@ function learn!(nniap, ds, opt::Flux.Optimise.AbstractOptimiser, epochs, loss, w
...
+    ds = ds_train |> _device
```
Comment (emmanuellujan) [inline examples/ACE/utils/NNIAP.jl:143]: This does not transfer the dataset to the GPU. Furthermore, the function `learn!` will transfer batches of training data to the GPU. For large data sets it is not possible to transfer all the training data to the GPU at once, which is one of the reasons why batches are used. Please, remove `|> _device`

File: examples/ACE/utils/NNIAP.jl [hunk 18]
```diff
@@ -72,12 +130,45 @@ function learn!(nniap, ds, opt::Flux.Optimise.AbstractOptimiser, epochs, loss, w
...
+    optim = Flux.setup(opt, nn) |> _device
+    ∇loss(nn, iap, batch, true_energy, local_descriptors, w_e, w_f) = gradient((nn) -> loss(nn, iap, batch, true_energy, local_descriptors, w_e, w_f), nn)
```
Comment (emmanuellujan) [inline examples/ACE/utils/NNIAP.jl:145]: Please, remove `batch` from the interface. 

It will probably be necessary to add a vector with the number of atoms per configuration of the `local_descriptors` batch array.

**Review by emmanuellujan — COMMENTED** [review]
I left a new battery of comments. The idea is that in this PR only GPU energy training is  implemented. In a next PR the same can be addressed for forces. The code is converging, I think a few more iterations are needed. Good job, keep up the good work 💪

**Review by emmanuellujan — COMMENTED** [review]
I left a new round of comments. We are gradually converging. Keep it up!

File: examples/ACE/fit-neural-ace.jl [hunk 19]
```diff
@@ -26,7 +25,7 @@ args = ["experiment_path",      "a-Hfo2-300K-NVT-6000-NeuralACE/",
         "n_test_sys",           "100",
         "n_red_desc",           "0", # No. of reduced descriptors. O: don't apply reduction
         "nn",                   "Chain(Dense(n_desc,8,relu),Dense(8,1))",
-        "n_epochs",             "100",
+        "n_epochs",             "5",
```
[the remark this answers is not in the document]
Comment (emmanuellujan) [reply examples/ACE/fit-neural-ace.jl:28]: This comment should be addressed

File: examples/ACE/fit-neural-ace.jl [hunk 20]
```diff
@@ -36,7 +35,13 @@ args = ["experiment_path",      "a-Hfo2-300K-NVT-6000-NeuralACE/",
...
         "csp",                  "1.0",
         "w_e",                  "0.01",
-        "w_f",                  "1.0"]
+        "w_f",                  "1.0",
+        "device",              "gpu"]
+
+        n_batches = 100
+
+        n_epochs = 5
+        n_batches = 1
```
[the remark this answers is not in the document]
Comment (emmanuellujan) [reply examples/ACE/fit-neural-ace.jl:44]: This comment should be addressed

File: examples/ACE/fit-neural-ace.jl [hunk 21]
```diff
@@ -36,7 +35,13 @@ args = ["experiment_path",      "a-Hfo2-300K-NVT-6000-NeuralACE/",
         "wL",                   "1.0",
         "csp",                  "1.0",
         "w_e",                  "0.01",
-        "w_f",                  "1.0"]
+        "w_f",                  "1.0",
+        "device",              "gpu"]
```
Comment (emmanuellujan) [inline examples/ACE/fit-neural-ace.jl:39]: The default should be cpu, because today we do not have a complete gpu implementation.

File: examples/ACE/fit-neural-ace.jl [hunk 22]
```diff
@@ -93,14 +95,23 @@ end
...
 n_epochs = input["n_epochs"]
-learn!(nace, ds_train, opt, n_epochs, loss, w_e, w_f)
+
+# learn!(nace |> cpu, ds_train |> cpu, opt |> cpu, n_epochs, loss, w_e, w_f)
+# learn!(nace, ds_train, opt, n_epochs, loss, w_e, w_f, cpu)
+n_batches = 3
+_device = gpu
+# learn!(nace, ds_train |> _device, opt |> _device, n_epochs, loss, w_e, w_f, 1.0, 1.0, _device, n_batches)
+learn!(nn |> _device, ace |> _device, ds_train |> _device, opt |> _device, n_epochs, loss, w_e, w_f, 1.0, 1.0, _device, n_batches)
+
```
[the remark this answers is not in the document]
Comment (emmanuellujan) [reply examples/ACE/fit-neural-ace.jl:119]: This comment should be addressed

**Review by emmanuellujan — COMMENTED** [review]
I think this first version of the gpu energy training looks good. There are some comments that still need to be addressed. Keep it up :)

**Review by emmanuellujan — APPROVED** [review]

**Comment by emmanuellujan** [comment]
As this PR is getting longer, and I need to organize ace-experiments, I will merge it, and then we can follow up.
